### PR TITLE
[Named min timestamp leases] Fix break in RemoteTimelockServiceAdapter

### PR DESCRIPTION
**Pull request by ergo14**
## General
**Before this PR**:
#7398 broke a public API used internally on two products.
**After this PR**:
Add a factory and publicize a method to make fixing the break tractable.
==COMMIT_MSG==
==COMMIT_MSG==

**Priority**:
P1
**Concerns / possible downsides (what feedback would you like?)**:
See PR comments
**Is documentation needed?**:
No
## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:
No, this is to help fix a break
**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:
No
**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:
Yes
**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:
No
**Does this PR need a schema migration?**
No
## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
N/A
**What was existing testing like? What have you done to improve it?**:
N/A
**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
N/A
**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:
N/A
## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:
N/A
**Has the safety of all log arguments been decided correctly?**:
No
**Will this change significantly affect our spending on metrics or logs?**:
No
**How would I tell that this PR does not work in production? (monitors, etc.)**:
N/A
**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:
N/A
**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:
N/A
## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
No
**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
No
**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
No
## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@raiju

### CODE REVIEW

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbDialogueServiceProvider.java [hunk 1]
```diff
@@ -123,7 +123,7 @@ public ConjureTimelockService getConjureTimelockService() {
         return new TimeoutSensitiveConjureTimelockService(shortAndLongTimeoutServices);
     }
 
-    MultiClientConjureTimelockServiceBlocking getMultiClientConjureTimelockService() {
+    public MultiClientConjureTimelockServiceBlocking getMultiClientConjureTimelockService() {
```
Comment (ergo14) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/AtlasDbDialogueServiceProvider.java:126]: There are three places where the break existed. For each one, they can either:
- Construct a conjure client using internal server lib
- Have an instance of `AtlasDbDialogueServiceProvider`

This is why I expose this here as public

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 2]
```diff
@@ -121,6 +130,28 @@ public static RemoteTimelockServiceAdapter create(
                 minLeasedTimestampGetter);
     }
 
+    public static RemoteTimelockServiceAdapter create(
+            Namespace namespace,
+            TimelockRpcClient baseRpcClient,
+            ConjureTimelockService baseConjureClient,
+            MultiClientConjureTimelockServiceBlocking multiClientConjureClient) {
```
Comment (ergo14) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:137]: I am trying to scope down the parameters provided here so we can later make some classes private (e.g. `NamespacedConjureTImelockServiceImpl`)

And, all users who were broken effectively do this

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 3]
```diff
@@ -121,6 +130,28 @@ public static RemoteTimelockServiceAdapter create(
                 minLeasedTimestampGetter);
     }
 
+    public static RemoteTimelockServiceAdapter create(
```
Comment (ergo14) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:133]: To keep in mind for the weirdness in this class https://github.com/palantir/atlasdb/pull/7398#discussion_r1824772592

As I said there, the public constructor + public factory situation existed before and I can track standardizing in a separate PR if you'd like that

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 4]
```diff
@@ -121,6 +130,28 @@ public static RemoteTimelockServiceAdapter create(
...
+            Namespace namespace,
+            TimelockRpcClient baseRpcClient,
+            ConjureTimelockService baseConjureClient,
+            MultiClientConjureTimelockServiceBlocking multiClientConjureClient) {
+        NamespacedTimestampLeaseService timestampLeaseService = new NamespacedTimestampLeaseServiceImpl(
+                namespace, new AuthenticatedInternalMultiClientConjureTimelockService(multiClientConjureClient));
```
Comment (jeremyk-91) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:139]: yeah was basically expecting to see something like this

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java (hunk 3 above)
Comment (jeremyk-91) [reply lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:133]: Ah, okay. Womp - wasn't aware of that - let's do that if feasible? Thanks!

**Review by jeremyk-91 — APPROVED** [review]
👍

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java (hunk 3 above)
Comment (ergo14) [reply lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:133]: Tracking internally!

**Pull request merged**